e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,216 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	monitoringv2 "cloud.google.com/go/monitoring/apiv3/v2"
+	"cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"google.golang.org/api/iterator"
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/retry"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/kinds"
+	ocmetrics "kpt.dev/configsync/pkg/metrics"
+)
+
+const (
+	DefaultMonitorKSA     = "default"
+	MonitorGSA            = "e2e-test-metric-writer"
+	GCMExportErrorCaption = "failed to export time series to GCM"
+	GCMMetricPrefix       = "custom.googleapis.com/opencensus/config_sync"
+)
+
+var GCMMetricTypes = []string{ocmetrics.ReconcilerErrors.Name(), ocmetrics.PipelineError.Name(), ocmetrics.ReconcileDuration.Name(), ocmetrics.ParserDuration.Name(), ocmetrics.InternalErrors.Name()}
+
+func TestOtelCollectorDeployment(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation1, ntopts.RequireGKE(t))
+	nt.T.Cleanup(func() {
+		nt.MustKubectl("delete", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml", "--ignore-not-found")
+		if t.Failed() {
+			nt.PodLogs("config-management-monitoring", ocmetrics.OtelCollectorName, "", false)
+		}
+	})
+
+	// If Workload Identity enabled on cluster, setup KSA to GSA annotation
+	if workloadPool, err := getWorkloadPool(nt); err != nil {
+		nt.T.Fatal(err)
+	} else if workloadPool != "" {
+		nt.T.Log("Workload identity enabled, adding KSA annotation to use %s service account", MonitorGSA)
+		nt.MustKubectl("annotate", "serviceaccount", "--namespace", ocmetrics.MonitoringNamespace, DefaultMonitorKSA,
+			fmt.Sprintf("iam.gke.io/gcp-service-account=%s@%s.iam.gserviceaccount.com", MonitorGSA, *e2e.GCPProject))
+	}
+
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap, log and IAM")
+	nomostest.DeletePodByLabel(nt, "app", ocmetrics.OpenTelemetry, false)
+	if err := watchForOtelCollectorStatusCurrent(nt); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Watch for metrics in GCM, timeout 2 minutes")
+	ctx := context.Background()
+	client, err := createGCMClient(ctx)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	// retry for 2 minutes until metric is accessible from GCM
+	_, err = retry.Retry(120*time.Second, func() error {
+		for _, metricType := range GCMMetricTypes {
+			descriptor := fmt.Sprintf("%s/%s", GCMMetricPrefix, metricType)
+			it := listMetricInGCM(ctx, nt, client, -2*time.Minute, descriptor)
```

Comment:
Is 2m always going to be enough?

Do we expect to see all the metrics in the list in this time frame if we didn't trigger any actions?

Reply:
From what I've seen so far (in freshtestenv) 2m is good enough, the initial commit should be syncing so no additional trigger actions are needed.  The metrics listed in GCMMetricTypes can all be detected once the exporter is up running.

Reply:
Instead of subtracting two minutes, could we record the time right after we restart the Pod and use that as the start time?

Reply:
Good point, just did, using the time right after otel-collector became current as startTime.